File: Graph-2-Ranking/Ranking.cpp
#include <bits/stdc++.h>
#define INF 1e9
#define WIN 1
#define LOSE -1

using namespace std;

int solution(int n, vector<vector<int>> results)
{
    int answer = 0;
    int graph[n + 1][n + 1];

    for (int i = 0; i <= n; i++)
    {
        fill(graph[i], graph[i] + (n + 1), INF);
        graph[i][i] = 0;
    }

    for (int i = 0; i < results.size(); i++)
    {
        graph[results[i][0]][results[i][1]] = WIN;
        graph[results[i][1]][results[i][0]] = LOSE;
    }

    for (int k = 1; k <= n; k++)
    {
        for (int i = 1; i <= n; i++)
        {
            for (int j = 1; j <= n; j++)
            {
                if (graph[i][k] == WIN && graph[k][j] == WIN)
                {
                    graph[i][j] = WIN;
                    graph[j][i] = LOSE;
                }
            }
        }
    }

    for (int i = 1; i <= n; i++)
    {
        int cnt = 0;
        for (int j = 1; j <= n; j++)
        {
            if (graph[i][j] == 1 || graph[i][j] == -1)
                cnt++;
        }
        if (cnt == n - 1)
            answer++;
    }

    return answer;
}

int main()
{
    int n = 5;
    vector<vector<int>> results = {{4, 3}, {4, 2}, {3, 2}, {1, 2}, {2, 5}};
    cout << solution(n, results) << endl;
    return 0;
}

File: Graph-2-Ranking/Ranking.py
from collections import defaultdict


def DFS(G, IO, V, i, BASE):
    V[i] = True
    if i != BASE:
        IO[BASE][1] += 1
        IO[i][0] += 1
    for x in G[i]:
        if not V[x]:
            DFS(G, IO, V, x, BASE)


def solution(n, results):
    G = [[] for _ in range(n+1)]
    IO = defaultdict(lambda: [0, 0])
    answer = 0

    for a, b in results:
        G[a].append(b)

    for i in range(1, n+1):
        DFS(G, IO, [False] * (n+1), i, i)

    for i in range(1, n+1):
        if IO[i][0] + IO[i][1] == n - 1:
            answer += 1

    return answer


print(solution(6, [[4, 3], [4, 2], [3, 2], [1, 2], [2, 5], [2, 6]]))
